cmd/update/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+    #[clap(long)]
+    /// Don't commit the image, only write
+    skip_commit: bool,
+    /// path to binary to write
+    path: PathBuf,
+}
+
+fn update(context: &mut humility::ExecutionContext) -> Result<()> {
+    let core = &mut **context.core.as_mut().unwrap();
+    let hubris = context.archive.as_ref().unwrap();
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+
+    let subargs = UpdateArgs::try_parse_from(subargs)?;
+    let mut context = HiffyContext::new(hubris, core, subargs.timeout)?;
+    let funcs = context.functions()?;
+
+    let start_update_fn = funcs.get("StartUpdate", 0)?;
+    let write_block_fn = funcs.get("WriteBlock", 2)?;
+    let finish_update_fn = funcs.get("FinishUpdate", 0)?;
+    let block_size_fn = funcs.get("BlockSize", 0)?;
+
+    let t = hubris.lookup_task("update_server").unwrap();
+    let m = hubris.lookup_module(*t)?;
+    let update_enum = m.lookup_enum_byname(hubris, "UpdateTarget")?;
+
+    let update_target = match update_enum.lookup_variant_byname(&subargs.target)
+    {
+        Ok(v) => v.tag.unwrap(),
+        _ => bail!("Couldn't find image target {}", subargs.target),
+    };
+
+    let ops = vec![Op::Call(block_size_fn.id), Op::Done];
+
+    let results = context.run(core, ops.as_slice(), None)?;
+
+    let blk_size = match &results[0] {
+        Ok(s) => {
+            if s.len() < 4 {
+                bail!("Bad length for block size!");
+            }
+            (s[0] as u32)
+                | (s[1] as u32) << 8
+                | (s[2] as u32) << 16
+                | (s[3] as u32) << 24
+        }
+        Err(_) => bail!("Failed to get block size"),
+    };
+
+    humility::msg!(
+        "Starting update using an update block size of {}",
+        blk_size
+    );
+    humility::msg!("(Erase may take a moment)");
+    let binary_contents = std::fs::read(&subargs.path)?;
+
+    let ops = vec![
+        Op::Push32(update_target as u32),
+        Op::Call(start_update_fn.id),
+        Op::Done,
+    ];
+
+    let _ = context.run(core, ops.as_slice(), None)?;
+
+    let bar = ProgressBar::new(binary_contents.len() as u64);
+    bar.set_style(ProgressStyle::default_bar().template(
+        "humility: writing update image [{bar:30}] {bytes}/{total_bytes}",
+    ));
+
+    for (i, c) in binary_contents.chunks(blk_size as usize).enumerate() {
+        bar.set_position((i * (blk_size as usize)) as u64);
+
+        let ops = vec![
+            Op::Push32(blk_size),
+            Op::Push32(i as u32),
+            Op::Call(write_block_fn.id),
+            Op::DropN(2),
+            Op::Done,
+        ];
+
+        let _ = context.run(core, ops.as_slice(), Some(c))?;
+    }
+
+    bar.finish_and_clear();
+    if subargs.skip_commit {
+        humility::msg!("Not committing update");
+    } else {
+        humility::msg!("Comitting update");
+        let ops = vec![Op::Call(finish_update_fn.id), Op::Done];
+        let _ = context.run(core, ops.as_slice(), None)?;
+    }
```

Comment:
For systems with auxiliary flash data (just Sidecar at the moment), that flash _also_ needs to be programmed for the system to be functional.

This is a bit awkward: programming the auxflash uses Idol APIs, so we have to reboot into the new image to access them (since we have no knowledge of the previous image).

[Here's](https://github.com/oxidecomputer/humility/blob/master/cmd/flash/src/lib.rs#L429-L506) where that's done in `humility flash`.

It looks like this `update` subcommand doesn't reboot into the new image at all (with or without `--skip-commit`), and also doesn't program the auxiliary flash, meaning you could end up with broken systems next time it's rebooted.

For now, I'd vote to either

(a) Refuse to run `humility update` if there is auxiliary flash data in the Hubris archive **at all**, or
(b) Refuse to run `humility update` if there's auxflash data **and** `--skip-commit` is set, and reboot into the new image + program the auxflash if `--skip-commit` is not set

(more generally, refuse to run a `humility update` command that could leave the system broken)

Reply:
I think this is expecting `humility update` to work at a much higher level than it currently does.  It has no knowledge of hubris archives and the path that it takes is a raw binary. The existing behavior is closer to `humility gpio` than `humility flash` in that it can in fact put the system in a weird state which is actually very useful for testing. I agree that what you're describing is what we will eventually want but I'm also not sure if this is the right time to start adding it since we're still working out other parts of the update sequence.

Reply:
Okay, that makes sense – if this is taking raw `.bin` files (rather than Hubris archives), then I agree that it shouldn't care about auxflash.